File: src/scripts/protocols/gamespy2.ts
import { InvalidResponseError } from '../lib/response-error'
import { DataReader } from '../lib/data-reader'
import { DataWriter } from '../lib/data-writer'
import { genRandomInteger } from '../lib/gen-random-integer'

export const info: ProtocolInfo = {
  id: 'gamespy2',
  name: 'GameSpy 2',
  feature: 'query',
  transport: 'udp',
}

let requestID = 0

export const query: QueryFn = () => {
  requestID = genRandomInteger(0, 256 ** 4)

  const data = new DataWriter()
    .u8(0xfe).u8(0xfd) // header
    .u8(0x00)          // delimiter
    .u32le(requestID)  // request id
    .u8(0xff)          // get server info and rules
    .u8(0xff)          // get player info
    .u8(0xff)          // get team info
    .buf

  gsw.send(data)
}

const normalizeServerInfo = (inf: Info): ServerInfo => ({
  [InfoField.SERVER_NAME]: String(inf.hostname),
  [InfoField.NUM_PLAYERS]: Number(inf.numplayers),
  [InfoField.MAX_PLAYERS]: Number(inf.maxplayers),
  [InfoField.MAP]: String(inf.maptitle || inf.mapname),
  [InfoField.GAME_VERSION]: String(inf.gamever),
  [InfoField.SECURE]: inf.sv_punkbuster === '1',
})

const parseServerInfo = (r: DataReader, delimiter: number): Info => {
  const inf: Info = {}

  while (!r.is_end) {
    const key = r.zstring(delimiter)
    const val = r.zstring(delimiter)

    if (!(key && val)) {
      break;
    }

    inf[key] = val
  }

  return inf
}

const readPlayerFields = (r: DataReader, delimiter: number) => {
  const fields: PlayerField[] = []

  while (!r.is_end) {
    const field = r.zstring(delimiter)

    if (!field) {
      break;
    }

    fields.push({
      title: field.slice(0, 1).toLocaleUpperCase() + field.slice(1),
      kind: 'string',
      field: field,
      main: field === 'name',
    })
  }

  return fields
}

const parsePlayerList = (r: DataReader, delimiter: number) => {
  const players: Player[] = []
  const player_count = r.u8()
  const pfields = readPlayerFields(r, delimiter)
  const field_count = pfields.length

  for (let iplayer = 0; iplayer < player_count && !r.is_end; iplayer++) {
    const player: Player = {}

    for (let ifield = 0; ifield < field_count && !r.is_end; ifield++) {
      const val = r.zstring(delimiter)
      player[pfields[ifield].field] = val
    }

    players.push(player)
  }

  return { pfields, players }
}

export const processResponse: ProcessResponseFn = data => {
  const r = new DataReader(data)

  const delimiter = r.u8()
  const reqid = r.u32le()

  if (reqid !== requestID) {
    throw new InvalidResponseError('invalid request id')
  }

  const details = parseServerInfo(r, delimiter)
  const { pfields, players } = parsePlayerList(r, delimiter)
  const inf = normalizeServerInfo(details)

  gsw.sinfo(details, inf)
  gsw.plist(pfields, players)
}

File: src/protocols/gamespy2.lua
local to_boolean = require("lib/to_boolean")
local DataWriter = require("lib/DataWriter")
local DataReader = require("lib/DataReader")

protocol = {
  id        = "gamespy2",
  name      = "GameSpy 2",
  feature   = "query",
  transport = "udp",
}

---@class Field
---@field title string
---@field kind string
---@field field string

---@type integer
local requestId = 0

function query()
  local w = DataWriter ()
    :u8(0xfe):u8(0xfd) -- header
    :u8(0x00)          -- delimiter
    :u32le(requestId)  -- request id
    :u8(0xff)          -- get server info and rules
    :u8(0xff)          -- get player info
    :u8(0xff)          -- get team info

  gsw.send(w.buf)
end

--- @param details table
local function normalize_server_info(details)
  return {
    server_name = details.hostname,
    num_players = tonumber(details.numplayers),
    max_players = tonumber(details.maxplayers),
    map = details.maptitle or details.mapname,
    game_version = details.gamever,
    game_mode = details.gametype or details.gamemode,
    secure = to_boolean(details.sv_punkbuster),
  }
end

--- @param r DataReader
--- @param delimiter integer
local function parse_server_info(r, delimiter)
  local details = {}

  while not r:is_end() do
    local key = r:zstring(delimiter)
    local val = r:zstring(delimiter)

    if #key == 0 and #val == 0 then
      break
    end

    details[key] = val
  end

  return details
end

--- @param r DataReader
--- @param delimiter integer
local function read_player_fields(r, delimiter)
  ---@type Field[]
  local fields = {}

  while not r:is_end() do
    local field = r:zstring(delimiter)

    if #field == 0 then
      break
    end

    table.insert(fields, {
      title = field:sub(1, 1):upper() .. field:sub(2, -2),
      kind = "string",
      field = field,
    })
  end

  return fields
end

--- @param r DataReader
--- @param delimiter integer
local function parse_player_list(r, delimiter)
  local players = {}
  local player_count = r:u8()
  local pfields = read_player_fields(r, delimiter)

  for player_idx = 1, player_count do
    if r:is_end() then
      break
    end

    local player = {}

    for field_idx = 1, #pfields do
      local val = r:zstring(delimiter)
      player[pfields[field_idx].field] = val
    end

    players[player_idx] = player
  end

  return pfields, players
end

---@param data string
function process(data)
  local r = DataReader(data)
  local delimiter = r:u8()
  local reqid = r:u32le()

  if reqid ~= requestId then
    error("InvalidResponseError: invalid request id")
  end

  local details = parse_server_info(r, delimiter)
  local pfields, players = parse_player_list(r, delimiter)
  local inf = normalize_server_info(details)

  gsw.sinfo(details, inf)
  gsw.plist(pfields, players)
end
